core/src/iam/entities/resources/level.rs

```diff
@@ -17,7 +17,7 @@ pub enum Level {
 	Root,
 	Namespace(String),
 	Database(String, String),
-	Scope(String, String, String),
+	Record(String, String, String),
```

Comment:
If we don't restrict it to a record we cannot assume to run the `.record()` method on the result of signin/signup clauses, making record auth a bit more tedious to work with. Could you come up with an example where we would want to store arbitrary data?
```
DEFINE ACCESS .. TYPE CUSTOM
  SIGNIN (
    (SELECT * FROM user WHERE bla).id
  );
```

Reply:
I think I understand what you mean. I pondered about this authorization level being associated with an access method instead of a record. In the end, I decided against it because I think in the current context, authorization levels refer to the logical level at which the authenticated actor is defined. I believe, even with a wider system of permissions, actors will still need to be defined somewhere specific in the key-value store. The record level identifies the case where the authenticated actor is created not at the database level (which requires database level permissions with owner role) but in the record level, which can be modified by any actor with record level authorization if allowed by permissions. I think this distinction is sufficient to warrant a record authorization level.

I imagine you are asking about considering the possibility of a more flexible access level which is not bound by any specific level, but by a set of explicitly defined boundaries. Even if that is the direction where we want to go, I don't think that would be dependent on the authorization level. This would just allow actors defined at a level (where the record level will still be a meaningful one) to act on resources that exist in a higher level if given permissions at that level by an actor who is authorized to do so. I think even in that case, the record authorization level is still meaningful to identify where the original actor was actually defined.

If an arbitrary object can be authenticated against and become an actor in the authorization context, this object will still exist at some of those four levels. If we ever add or consider an additional logical level where actors can live (maybe table could be one), we can explicitly add it to the enumerator.

Reply:
So my thoughts were:

```sql
DEFINE ACCESS ...
    SIGNIN AS (
        IF $session.ip = '123.456.789.012' {
            RETURN {
                type: 'admin',
                settings: {
                    people: 'crud',
                    deals: 'crud',
                    tasks: 'crud',
                }
            }
        } else {
            RETURN SELECT *, 'normal' AS type 
                FROM user 
                WHERE email = $user 
                FETCH settings
        }
    )
```

Reply:
Okay, I see what you mean. I think it would make sense to be able to offer something like this. What do you think about keeping the `Record` level, which I think makes sense on its own, and including the `data` property and functions to access any data associated with the authenticated actor. For now, this will only contain the record data whenever there is record-level access. Then, we leave implementation of this `Custom` access level for a separate PR, together with the functionality that will make this access possible and relevant. Since this will be adding a variant to the end of the enum and access to this data will be done through the already existing property and function, this change will be backward compatible.